user units: implicitly enable PrivateUsers= when sandboxing options are set

bluca opened the pull request:
Enabling these options when not running as root requires a user namespace, so implicitly enable PrivateUsers=.

bluca commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
+                                s->exec_context.n_bind_mounts > 0 ||
+                                s->exec_context.n_temporary_filesystems > 0 ||
+                                s->exec_context.root_directory ||
+                                s->exec_context.extension_directories ||
```
Up until this one the unit layout changes, so definitely should be enabled. For the ones below it only affects what they can touch though, so less sure.

poettering replied:
yeah, everything after this should be dropped.

yuwata commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
 
+        /* These options require PrivateUsers= when used in user units, as we need to be in a user namespace
+         * to have permission to enable them when not running as root. */
+        if (!s->exec_context.private_users && !MANAGER_IS_SYSTEM(UNIT(s)->manager))
```
Is this only for service units? Why?

bluca replied:
It's for `[Service]` sandboxing options? Where else would it be used?

yuwata commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
+                                s->exec_context.read_only_paths ||
+                                s->exec_context.inaccessible_paths ||
+                                s->exec_context.exec_paths ||
+                                s->exec_context.no_exec_paths;
```
Here, `s->exec_runtime` (for `NetworkNamespacePath=` and `IPCNamespacePath=`) and `s->exec_context.mount_flags` (for `MountFlags=`) need to be checked. At least to make consistent with the above man page update.

If these settings are not supported in user session. Please update the man page.

bluca replied:
Those are implied with the network namespace/ipc namespace settings no?

poettering commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
+        /* These options require PrivateUsers= when used in user units, as we need to be in a user namespace
+         * to have permission to enable them when not running as root. */
+        if (!s->exec_context.private_users && !MANAGER_IS_SYSTEM(UNIT(s)->manager))
+                s->exec_context.private_users =
```
please don't patch around in ExecContext here. we usually don't change loaded configuration, and expose the actually loaded stuff on the bus. if you'd change the setting here, then before the first time the the service is started we'll report one value here, and afterwards another.

instead in src/core/execute.c, look for the line:

```c
if (needs_sandboxing && context->private_users && !have_effective_cap(CAP_SYS_ADMIN)) {
```

and change that to also do the private user dance if the other options are set.

maybe add a helper accessor function:

```c
static bool exec_context_get_private_users(const ExecContext *c) {
       return c->private_users || c->n_bind_mounts > 0 || …;
}
```

and so on. and then change the aforementioned line to:

```c
if (needs_sandboxing && context_get_private_users(context) && !have_effective_cap(CAP_SYS_ADMIN)) {
```

you get the idea

poettering commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
+                s->exec_context.private_users =
+                                s->exec_context.private_tmp ||
+                                s->exec_context.private_devices ||
+                                s->exec_context.private_network ||
```
hmm?

bluca replied:
What about it? You need a user namespace to set up a network namespace when !root

poettering commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
+                                s->exec_context.private_ipc ||
+                                s->exec_context.private_mounts ||
+                                s->exec_context.mount_apivfs ||
+                                s->exec_context.dynamic_user ||
```
dynamic_user doesn't really make sense for --user instances

poettering commented on src/core/service.c:
```diff
@@ -1532,6 +1532,35 @@ static int service_spawn_internal(
+                                s->exec_context.private_network ||
+                                s->exec_context.private_ipc ||
+                                s->exec_context.private_mounts ||
+                                s->exec_context.mount_apivfs ||
```
all of the above should degrade "gracefully" if we can't apply them, hence we should *not* imply PrivateUsers= for them

bluca replied:
Which ones?

bluca replied:
Actually, I'm beginning to question this - why shouldn't all of these be made to work? What's the advantage of requiring extra hoops jumping for some options and not others? It would cause more confusion rather than less

poettering replied:
there are two kind of namespacing options:

1. sandboxing: these will hide things, make things unaccessible but not rearrange things. We generally make these options fail gracefully: if we cannot apply them due to lack of perms, lack of kernel support, running in a container or so, we'll simply not apply them. This should not break anything, it will just weaken protections.

2. options that rearrange stuff, such as BindPaths=, RootDirectory= and things. These options will rearrange the tree hiearchy, stuff that would normally appear at location X will appear at location Y, it's actually moved around, not just hidden or made unaccessible. If we can't apply these options, we'll fail starting the service.

I think for the second type of options it is fine to imply PrivateUsers=, if we don't the service will just fail. Hence there isn't really a compat break: making stuff work that previously didn't work is always OK. making stuff work differently that previously already worked in some other way would be a compat break.

For the first type of options otoh I think we should not imply PrivateUsers=. Because we already handle this gracefully: it's entirely OK to not apply them, and people could rely on that and have their stuff work. if we suddenly turn on privateUsers= for them, then things might end up working differently than before, and I'd call that a compat break.

bluca replied:
> For the first type of options otoh I think we should not imply PrivateUsers=. Because we already handle this gracefully: it's entirely OK to not apply them, and people could rely on that and have their stuff work. if we suddenly turn on privateUsers= for them, then things might end up working differently than before, and I'd call that a compat break.

But they always have to be explicitly requested? If you set "ReadOnlyPath=/foo", and /foo is read-only, then it's all working as intended? Also the line is not as clear - making a subtree inaccessible is not that different from bind mounting something over it, the filesystem view changes. Same for network namespaces, the network stack view changes, and so on. From the "compat" angle, this is a slight change in behaviour regardless, for all of these options. Might as well change them all, so that they make sense and just work as they should?

poettering replied:
PrivateUsers=yes means pretty much all users of the system (with the exception of root and the service's own user) will show up as "nobody", that's quite a change. We should not do that willy-nilly and pretend it was not a compat break...

hiding stuff and locking down access to stuff is systematically different from moving things to other places.

InaccessiblePath= and ReadOnlyPath= in a way only encode a contract: "hey, I won't need this stuff" or "hey, I am not gonna modify this". If systemd then follows up with making this read-only or making it disappear is not gonna change things as long as the contract is fulfilled by the service. Hence a graceful fallback makes sense, and has always been implemented.

But BindPaths= is systematically different: the contract sugegsts "hey, I expect to get resource X when I access path Y". we cannot just skip it ever, we have to fail if we can't make that happen, because the contract in this case implies "i am gonna use this" and not the opposite...

So to summarize:

1. Status quo: If ReadOnlyPath= cannot be executed via namespaces due to perms/containers/lack of kernel support we'll simply skip it, and nothing bad happens, "ls -al /home" will still look nice, with all home dirs owned by the users they belong to.

2. with you patch: If readonlypath= might not be executable, we'll allocate a userns with a mapping that hides all other users, and then make the specified dir read-only. But suddenly "ls -al /home" will look all bonkers, because everything is onwed by "nobody", and noone asked for it.

bluca replied:
A behaviour change is a behaviour change. The important point is, what's the cost vs benefit? On my Debian machine there are 0 references to ReadOnly/Protect/Inaccessible and so on under /usr/lib/systemd/user/. Do you see any on Fedora?
Doing the split you suggest is not cost-free either: it means keeping this weird split forever, where half the sandboxing options require an explicit unrelated configuration to be enabled, and half don't. Most of those always worked with PrivateUsers=yes, but nobody ever used them in user units, because nobody was aware they could work, because the setup is very confusing. Fixing this is a big win.

bluca replied:
By installing every package that provides a user unit I see the following:

- no Read/Write/Inaccessible path set anywhere
- no BindPath
- no RootDirectory/ExtensionDirectory
- one service using Protect* and Private* but it also already sets PrivateUsers (vvmd.service)
- one service using PrivateTmp without PrivateUsers (stunnel@.service)
- 4 services using ProtectSystem: mpd.service mpdscribble.service fluidsynth.service onedrive.service

So there's 5 units out of 284 that would be affected. I can probably get them fixed in a couple of days if needed, although looking at the comments I suspect they'd welcome it if this starts to work out of the box.

bluca replied:
I have created PRs for those 5 projects:

https://github.com/MusicPlayerDaemon/MPD/pull/1661
https://github.com/MusicPlayerDaemon/mpdscribble/pull/49
https://github.com/FluidSynth/fluidsynth/pull/1181
https://github.com/abraunegg/onedrive/pull/2205
https://salsa.debian.org/debian/stunnel/-/merge_requests/2 (downstream-only unit)

bluca replied:
All of the above have been fixed.
Found 2 more that are only shipped in Fedora:

https://gitlab.freedesktop.org/benzea/uresourced/-/merge_requests/24
https://github.com/abrt/reportd/pull/11

bluca replied:
Both of those were merged too

Werkov replied:
Ack, they're stronger protections (against leg shooting).

Back to the `Private*` topic, I might have misunderstood inheritable&ambient capabilities (or they're initialized differently) because I tried the following (6.1.0-rc7) and it works:

```
$ unshare -U --keep-caps bash     # this is user@foo.service, PrivateUsers=yes
$ unshare -m bash                 # forked child (user service) creates mount ns
$ mount --bind /home /home -oro   # can set up protections as I wish, caps survive exec
```

I.e. it's not necessary to create userns in each user service but just run `user@.service` in a user namespace.

bluca replied:
I don't think we can enable a user namespace for the whole user session, too many things might break, as any and all programs ran by a user will be affected

poettering commented on src/core/execute.c:
```diff
@@ -4667,7 +4704,7 @@ static int exec_child(
         }
 
-        if (needs_sandboxing && context->private_users && !have_effective_cap(CAP_SYS_ADMIN)) {
+        if (needs_sandboxing && exec_context_need_unprivileged_private_users(context)) {
```
"unprivileged"? why that? if this is applied to system services, there is no "unprivileged" involved?

poettering commented:
Would a new `PrivateUser=identity` concept work for this too? i.e. a userns that is private to the service, but just a full identity mapping?

bluca commented:
> Would a new `PrivateUser=identity` concept work for this too? i.e. a userns that is private to the service, but just a full identity mapping?

It turns out it cannot be done from the user manager, as it requires privileges to set up an id map of more than the current uid

DaanDeMeyer commented:
> > Would a new `PrivateUser=identity` concept work for this too? i.e. a userns that is private to the service, but just a full identity mapping?
> 
> It turns out it cannot be done from the user manager, as it requires privileges to set up an id map of more than the current uid

@brauner and @poettering had the idea to add a service to manage uid/gid ranges that would hand out user namespaces with a uid/gid range and idmapping configured to tools and services. Rootless nspawn would make use of this. I wonder if we could use the same service here to get an identity mapping without needing privileges? Not sure about the security implications though.

bluca commented:
> > > Would a new `PrivateUser=identity` concept work for this too? i.e. a userns that is private to the service, but just a full identity mapping?
> > 
> > 
> > It turns out it cannot be done from the user manager, as it requires privileges to set up an id map of more than the current uid
> 
> @brauner and @poettering had the idea to add a service to manage uid/gid ranges that would hand out user namespaces with a uid/gid range and idmapping configured to tools and services. Rootless nspawn would make use of this. I wonder if we could use the same service here to get an identity mapping without needing privileges? Not sure about the security implications though.

If it worked it could be retrofitted to use it yes - but it can be done later

bluca commented:
> > > Would a new `PrivateUser=identity` concept work for this too? i.e. a userns that is private to the service, but just a full identity mapping?
> > 
> > 
> > It turns out it cannot be done from the user manager, as it requires privileges to set up an id map of more than the current uid
> 
> @brauner and @poettering had the idea to add a service to manage uid/gid ranges that would hand out user namespaces with a uid/gid range and idmapping configured to tools and services. Rootless nspawn would make use of this. I wonder if we could use the same service here to get an identity mapping without needing privileges? Not sure about the security implications though.

mmh I tried this, but when the receiving end of the namespace FD is not the same uid, I get -EPERM back from `setns(fd, CLONE_NEWUSER)`, any idea what else needs to be done?

brauner commented:
> > > > Would a new `PrivateUser=identity` concept work for this too? i.e. a userns that is private to the service, but just a full identity mapping?
> > > 
> > > 
> > > It turns out it cannot be done from the user manager, as it requires privileges to set up an id map of more than the current uid
> > 
> > 
> > @brauner and @poettering had the idea to add a service to manage uid/gid ranges that would hand out user namespaces with a uid/gid range and idmapping configured to tools and services. Rootless nspawn would make use of this. I wonder if we could use the same service here to get an identity mapping without needing privileges? Not sure about the security implications though.
> 
> mmh I tried this, but when the receiving end of the namespace FD is not the same uid, I get -EPERM back from `setns(fd, CLONE_NEWUSER)`, any idea what else needs to be done?

Just summarizing what I explained elsewhere just now. To make this work you should have the `acquire_userns()` function setuid to the client's uid that requested the userns creation, then create the userns, and then have the parent write the mapping that you want. This ensures that the client owns the userns. It's also important that the client and the creator of the userns are located in the same userns - the actual restriction is a bit more detailed but it doesn't matter here - so if you do nested userns creation then you should take the client's userns into account so the parent userns relationship is correctly set up. Also, you should be ON HOLIDAY. STEP AWAY FROM THE COMPUTER.

bluca commented:
> Also, you should be ON HOLIDAY. STEP AWAY FROM THE COMPUTER.

That's why I have time to do the cool stuff! Beer-fueled hacking go brrrr

bluca commented:
Ok, seems to work as expected, PTAL

brauner commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,40 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner != UID_INVALID) {
+                        (void) setuid(owner);
+                        (void) unshare(CLONE_NEWUSER);
```
Why do you ignore return values here? Both can legitimately fail. I would check for failures in the child. Also, the freeze thing seems strange and is that even guaranteed to be correctly ordered such that the parent is guaranteed to try to write that mapping after the userns is unshared?

I would redo this and align this with how this is more commonly done:
* fork child and use socketpair()
* blocking read in parent
* let child to setuid and create userns. then write to socket to let parent know that we're ready for it to write mapping
* read from socket succeeds and child told us that we can write the mapping
* write the idmapping for the child and open the namespace file descriptor
* inform child that we're done writing the mapping and close socket
* wait for child to exit

This is way cleaner and guarantees correct ordering under all circumstances.

bluca replied:
the freeze stuff is pre-existing, and closing the pipe is the same as writing to a socket and it's how it's done elsewhere

bluca replied:
> Why do you ignore return values here? Both can legitimately fail. I would check for failures in the child

Follow-up operations will fail if those don't work, so there's no need for extra complications with transferring error codes over

bluca replied:
The resulting namespace cannot be used by the caller, joining it will fail with permission denied, I'm pretty sure

brauner commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,40 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
-         * process. We can use that for file system mounts and similar. */
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure the mapping is done after setuid(). */
```
That comment is wrong. The mapping isn't relevant. What's relevant is that the userns is created as the user that's supposed to own it.

bluca replied:
doesn't the mapping have to be set up _after_ the setuid?

brauner replied:
No, the mapping can be written whenever the userns has been created. As soon as a process has called unshare() or forked into a new userns you can right away write the mapping.

bluca replied:
right, the order has to be setuid -> unshare -> map, got confused and forgot to document the middle part

brauner commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,40 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+        /* If we have to set a different owner we have to setuid before unsharing the new namespace, otherwise
+         * the kernel will refuse the operation. */
+        if (owner == UID_INVALID)
+                flags |= FORK_NEW_USERNS;
```
I'm sorry I understand this less and less. Why do you raise `FORK_NEW_USERNS` when the uid is invalid? The explanation on top doesn't make sense to me.

brauner replied:
Ok, I see. It's for the case where we don't have to have the userns owned by a specific uid. Sorry, I missed that part

brauner replied:
Though I'd rephrase that comment. As it stands the code documents the code in the child setuid()+unshare(CLONE_NEWUSER) and not the code that it's on top of.

brauner commented on src/basic/namespace-util.c:
```diff
@@ -199,8 +200,10 @@ int detach_mount_namespace(void) {
 }
 
-int userns_acquire(const char *uid_map, const char *gid_map) {
+int userns_acquire(const char *uid_map, const char *gid_map, uid_t owner) {;
```
What's the semicolon doing here?

bluca replied:
santa left it there on xmas eve because we have been such good boys

brauner reviewed (changes requested):
I think having two different codepaths to create a user namespace in `acquire_userns()` isn't necessary and makes the function more difficult to understand than it needs to be. I would just consolidate this into something (untested) like:

```diff
diff --git a/src/basic/namespace-util.c b/src/basic/namespace-util.c
index a87a875943..8920616462 100644
--- a/src/basic/namespace-util.c
+++ b/src/basic/namespace-util.c
@@ -199,25 +199,63 @@ int detach_mount_namespace(void) {
         return RET_NERRNO(mount(NULL, "/", NULL, MS_SLAVE | MS_REC, NULL));
 }

-int userns_acquire(const char *uid_map, const char *gid_map) {
+int userns_acquire(const char *uid_map, const char *gid_map, uid_t owner) {
         char path[STRLEN("/proc//uid_map") + DECIMAL_STR_MAX(pid_t) + 1];
+        _cleanup_close_pair_ int socket_fd[2] = PIPE_EBADF;
         _cleanup_(sigkill_waitp) pid_t pid = 0;
-        _cleanup_close_ int userns_fd = -1;
+        _cleanup_close_ int userns_fd = -EBADF;
         int r;
+        char c = '1';

         assert(uid_map);
         assert(gid_map);

         /* Forks off a process in a new userns, configures the specified uidmap/gidmap, acquires an fd to it,
          * and then kills the process again. This way we have a userns fd that is not bound to any
-         * process. We can use that for file system mounts and similar. */
+         * process. We can use that for file system mounts and similar.
+         * If the owner of the new namespace is specified, setuid to it in the child. Use a pipe to sync and
+         * ensure the mapping is done after setuid(). */

-        r = safe_fork("(sd-mkuserns)", FORK_CLOSE_ALL_FDS|FORK_DEATHSIG|FORK_NEW_USERNS, &pid);
+        if (socketpair(PF_LOCAL, SOCK_STREAM | SOCK_CLOEXEC, 0, socket_fd) < 0)
+                return log_error_errno(errno, "Failed to open socketpair: %m");
+
+        r = safe_fork_full("(sd-mkuserns)", (int[]) { pipe_fd[1] }, 1, FORK_CLOSE_ALL_FDS|FORK_DEATHSIG, &pid);
         if (r < 0)
                 return r;
-        if (r == 0)
-                /* Child. We do nothing here, just freeze until somebody kills us. */
-                freeze();
+        if (r == 0) {
+                socket_fd[1] = safe_close(socket_fd[1]);
+
+                if (owner != UID_INVALID && setuid(owner))
+                        _exit(EXIT_FAILURE);
+
+                /* Create a new userns. It'll either be owned by @owner if it's
+                 * specified or by the caller's euid.
+                 *
+                 * Note, this userns has a full set of capabilities. IOW, the
+                 * user of that userns is responsible for dropping capabilities
+                 * and so on. */
+                ret = unshare(CLONE_NEWUSER);
+                if (ret)
+                        _exit(EXIT_FAILURE);
+
+                /* We're ready for the idmapping to be written. */
+                ret = write(sock_fds[0], &c, 1);
+                if (ret != 1)
+                        _exit(EXIT_FAILURE);
+
+                /* Stick around so the parent can persist our userns. */
+                ret = read(sock_fds[0], &c, 1);
+                if (ret != 1)
+                        _exit(EXIT_FAILURE);
+
+                _exit(EXIT_SUCCESS);
+        }
+
+        socket_fd[0] = safe_close(socket_fd[0]);
+
+        ret = read(sock_fds[1], &c, 1);
+        if (ret != 1)
+                return log_error_errno(r, "Failed to read from socketpair: %m");

         xsprintf(path, "/proc/" PID_FMT "/uid_map", pid);
         r = write_string_file(path, uid_map, WRITE_STRING_FILE_DISABLE_BUFFER);
@@ -233,6 +271,10 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
         if (r < 0)
                 return log_error_errno(r, "Failed to open userns fd: %m");

+        /* We don't care about the child anymore. Let it go away or get killed.
+         * Whatever happens first. */
+        (void)write(sock_fds[1], &c, 1);
+
         return TAKE_FD(userns_fd);

 }
```

bluca commented:
I want to use the atomic clone into new ns (`rawclone(CLONE_NEWUSER)`) as much as possible, hence I don't want to do that, given the setuid is only needed in one specific use case and the simpler path works for the rest

brauner commented:
> I want to use the atomic clone into new ns (`rawclone(CLONE_NEWUSER)`) as much as possible, hence I don't want to do that, given the setuid is only needed in one specific use case and the simpler path works for the rest

Why?

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,42 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner != UID_INVALID) {
+                        (void) setuid(owner);
+                        (void) unshare(CLONE_NEWUSER);
```
hmm, error check missing?

poettering replied:
also, i think we should bubble up the error code. i.e. write the errno you got into the pipe, read it in the parent

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,42 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+                        return -ETIMEDOUT;
+        }
 
         xsprintf(path, "/proc/" PID_FMT "/uid_map", pid);
```
if the setuid() failed for some reason (RLIMIT_NPROC for example) then you might end up writing into the uid_map of the originating userns. that looks very wrong.

plesae alway check for errors.

bluca replied:
The originating ns is the root one, so I don't think that can actually happen? But added anyway

poettering replied:
could happen in an nspawn container that has userns applied, no?

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,42 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+        if (r == 0) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner != UID_INVALID) {
+                        (void) setuid(owner);
```
hmm, shouldn't this be setresuid()?

setuid() is in 95% of cases not what you want... and if it actually is what you want, then we should document that here with a clear comment

poettering commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -1495,6 +1498,13 @@ node /org/freedesktop/systemd1 {
+      own user. This method sets up a new namespace, owned by the caller's uid, and returns a file descriptor
+      that the caller can change into. The caller will be root in the new namespace after changing into it,
+      so it should setuid back to its own user if it needs to run as such. Note that due to the security
+      implications, this method is gated by a new Polkit authorization policy (or CAP_SYS_ADMIN).</para>
```
hmm, I'd call this an "identity" mapping, no?

poettering commented on src/core/dbus-manager.c:
```diff
@@ -2783,6 +2783,50 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
+                return r;
+
+        if (asprintf(&line, UID_FMT " " UID_FMT " " UID_FMT "\n", 0u, 0u, DYNAMIC_UID_MIN - 1u) < 0)
+                return log_oom_debug();
```
hmm, so if this is used inside of a userns with a restricted uid range, this will fail, no?

i.e. we should probably read the current mapping, and not try to map more than what is currently mapped for ourselves.

poettering commented on src/core/dbus-manager.c:
```diff
@@ -2783,6 +2783,50 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
+        if (r == 0)
+                return 1; /* No authorization for now, but the async polkit stuff will call us again when it has it */
+
+        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_UID|SD_BUS_CREDS_AUGMENT, &creds);
```
drop SD_BUS_CREDS_AUGMENT. that's the flag you specify if you want stuff to be added from /proc/. But you don#t wnat that here, and it's not suitable for security decisions like yours here.

bluca replied:
I need the pid, and for that SD_BUS_CREDS_AUGMENT is needed, I do not get it without this

poettering replied:
if you want the PID use SD_BUS_CREDS_PID?

or do you mean to say UID? if so, does it give you the EUID without augment?

poettering commented on src/core/dbus-manager.c:
```diff
@@ -2783,10 +2783,47 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
+        assert(manager);
+        assert(ret_user);
+
+        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_PID|SD_BUS_CREDS_UID|SD_BUS_CREDS_AUGMENT, &creds);
```
drop augment

poettering commented on src/core/execute.c:
```diff
@@ -5145,6 +5226,24 @@ static int exec_child(
+         * move into the namespace. The point of having PID1 set us a mapping for us is to make sure there's no
+         * visible difference for the units due to the implicit user namespace. */
+        if (!MANAGER_IS_SYSTEM(unit->manager) && userns_set_up && !context->private_users) {
+                r = setgid(saved_gid);
```
setresgid(), no?

poettering commented on src/core/execute.c:
```diff
@@ -5145,6 +5226,24 @@ static int exec_child(
+                        return log_unit_error_errno(unit, errno, "Failed to set GID back to " GID_FMT ": %m", saved_gid);
+                }
+
+                r = setuid(saved_uid);
```
setresuid()

poettering commented on src/core/execute.c:
```diff
@@ -5145,6 +5226,24 @@ static int exec_child(
+                if (r < 0) {
+                        *exit_status = EXIT_USER;
+                        return log_unit_error_errno(unit, errno, "Failed to set UID back to " UID_FMT ": %m", saved_uid);
+                }
```
hmm, i don't follow here. in the classic code paths we should have done this too, no?

i..e in the status quo ante if we use PrivateUsers= we currently run as root inside the userns? we really shouldn, we should be the service user there too...

bluca replied:
In the status quo it depends on the `User=` property, as expected, because the userns comes later

poettering replied:
hmm, that is a bit icky no? we probably should ensure that if we open a userns we *always* do at least one setresuid() afterwards.

poettering commented:
hmm, here's a question: maybe the IPC API should not return a newly minted usernsfd, but instead just take a pidfd and write the identity onto it? (just asking, not sure myself)

poettering commented:
anyway, i am pretty sure the error checking in acquire_userns() is absolutely necessary, and we should bubble up the original error really. It's not clear to me at all that it would be safe to ignore those errors, i.e. that is guaranteed that the later syscalls will fail if the earlier ones failed. And even if that is guaranteed now, it's not clear that it will be guaranteed forever, i.e. people tend to start with locked down policies, and then allow more and more stuff as they go.

poettering commented:
> hmm, here's a question: maybe the IPC API should not return a newly minted usernsfd, but instead just take a pidfd and write the identity onto it? (just asking, not sure myself)

Passing fds from unpriv clients to priv IPC servers is always icky (since on Linux closing bad fds might hang). Hence I figure the current approach of returning a freshly allocated userns server-side and returning that sounds OK to me.

(though I am sure the audit people will hate you for that. But maybe let's not mention that ;-))

poettering commented on src/core/execute.c:
```diff
@@ -5145,6 +5226,24 @@ static int exec_child(
+         * after we are done setting up we want to go back to the original UID/GID, given we'll be root when we
+         * move into the namespace. The point of having PID1 set us a mapping for us is to make sure there's no
+         * visible difference for the units due to the implicit user namespace. */
+        if (!MANAGER_IS_SYSTEM(unit->manager) && userns_set_up && !context->private_users) {
```
instead of basically duplicating part of the check above, please set a new boolean var up top and check that here. i.e. `set_gid_uid_again = true` up top, or so.

these checks tend to be updated sooner or later, and thus if somebody accidentally updates one but not the other bad things will happen.

poettering commented:
and i'd really to understand why we didn't do the final setuid()/setgid() call you added now, before? i think we should always do that if we open a userns, regardles sif we do that via ipc or locally

bluca commented on src/core/dbus-manager.c:
```diff
@@ -2783,6 +2783,50 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
+                return sd_bus_error_setf(error, SD_BUS_ERROR_NOT_SUPPORTED,
+                                         "This method is only supported on the system manager.");
+
+        r = mac_selinux_access_check(message, "acquire-namespace", error);
```
I don't see an existing one that maps cleanly, and given we are bypassing the default security policy of the kernel here I figure a new one is better, so that existing services don't inherit the new capability unintentionally

bluca commented:
> It's not clear to me at all that it would be safe to ignore those errors, i.e. that is guaranteed that the later syscalls will fail if the earlier ones failed. 

They will fail. This API does not work without those calls, that's why I added them... if the child tmp process is not setuid'd then the caller cannot use it (unless the caller is root, but then it would not use that API in the first place, and can do whatever anyway). And you cannot alter the mapping of the root namespace.

But anyway, added the checks now even if they are not really useful, so it doesn't matter.

> and i'd really to understand why we didn't do the final setuid()/setgid() call you added now, before? i think we should always do that if we open a userns, regardles sif we do that via ipc or locally

I think it's because the userns comes later, and the user depends on User=/DynamicUser= and is otherwise root. So everything was always set up correctly from the get-go. But here we are switching things around, and doing the user namespace first.

brauner commented:
On Wed, Jan 04, 2023 at 05:01:25PM -0800, Luca Boccassi wrote:
> > It's not clear to me at all that it would be safe to ignore those errors, i.e. that is guaranteed that the later syscalls will fail if the earlier ones failed. 
> 
> They will fail.

They will fail in this usage. The idea of the API we proposed was always
to make it usable generically. And this is just not a safe general
assumption. We should never create a user namespace whose ownership is
undetermined until you try to setns() to it especially if we can
trivially avoid it.

DaanDeMeyer commented:
Unrelated to this PR but if we're going to require usage of this API for rootless nspawn it's going to be a pain since users will have to be running latest pid1 on the host to be able to use rootless nspawn. Building nspawn and repart and such from source is OK but having to run latest pid1 is a bit more complex.

bluca commented:
> On Wed, Jan 04, 2023 at 05:01:25PM -0800, Luca Boccassi wrote: > It's not clear to me at all that it would be safe to ignore those errors, i.e. that is guaranteed that the later syscalls will fail if the earlier ones failed. They will fail.
> They will fail in this usage. The idea of the API we proposed was always to make it usable generically. And this is just not a safe general assumption. We should never create a user namespace whose ownership is undetermined until you try to setns() to it especially if we can trivially avoid it.

Well sure, but I only tested this with the current use case, I cannot test it with future ones that don't exist yet, so I cannot guarantee it will work out of the box. It is possible more changes will be needed when adding new use cases like that, but that's generally an ok thing? It happens all the time...

But anyway it's all academic, I've applied all requested changes (apart from the dbus creds one, as it doesn't work without it as far as I can tell) and fixed an issue with GIDs found by the unit tests, PTAL.

brauner commented:
On Thu, Jan 05, 2023 at 02:49:12AM -0800, Daan De Meyer wrote:
> Unrelated to this PR but if we're going to require usage of this API
> for rootless nspawn it's going to be a pain since users will have to
> be running latest pid1 on the host to be able to use rootless nspawn.
> Building nspawn and repart and such from source is OK but having to
> run latest pid1 is a bit more complex.

I personally think that requirement is ok.

But if that's really something that's crucial to avoid we could do what
I originally also proposed: make it a separate very very minimal
namespace daemon binary that can be built alongside systemd-nspawn.

bluca commented:
> On Thu, Jan 05, 2023 at 02:49:12AM -0800, Daan De Meyer wrote: Unrelated to this PR but if we're going to require usage of this API for rootless nspawn it's going to be a pain since users will have to be running latest pid1 on the host to be able to use rootless nspawn. Building nspawn and repart and such from source is OK but having to run latest pid1 is a bit more complex.
> I personally think that requirement is ok. But if that's really something that's crucial to avoid we could do what I originally also proposed: make it a separate very very minimal namespace daemon binary that can be built alongside systemd-nspawn.

If we need systemd-backdoord.service we can always add it later and move the endpoint before the next release, this PR is already changing a fair few things, I don't really want to further expand it

brauner commented:
On Thu, Jan 05, 2023 at 07:10:56AM -0800, Luca Boccassi wrote:
> > On Thu, Jan 05, 2023 at 02:49:12AM -0800, Daan De Meyer wrote: Unrelated to this PR but if we're going to require usage of this API for rootless nspawn it's going to be a pain since users will have to be running latest pid1 on the host to be able to use rootless nspawn. Building nspawn and repart and such from source is OK but having to run latest pid1 is a bit more complex.
> > I personally think that requirement is ok. But if that's really something that's crucial to avoid we could do what I originally also proposed: make it a separate very very minimal namespace daemon binary that can be built alongside systemd-nspawn.
> 
> If we need systemd-backdoord.service we can always add it later and

Says the dude who wants to send around user namespaces with indeterminate ownership...;)
/me ducks

brauner commented:
> > On Thu, Jan 05, 2023 at 02:49:12AM -0800, Daan De Meyer wrote: Unrelated to this PR but if we're going to require usage of this API for rootless nspawn it's going to be a pain since users will have to be running latest pid1 on the host to be able to use rootless nspawn. Building nspawn and repart and such from source is OK but having to run latest pid1 is a bit more complex.
> > I personally think that requirement is ok. But if that's really something that's crucial to avoid we could do what I originally also proposed: make it a separate very very minimal namespace daemon binary that can be built alongside systemd-nspawn.
> 
> If we need systemd-backdoord.service we can always add it later and move the endpoint before the next release, this PR is already changing a fair few things, I don't really want to further expand it

That comment is weird. Nobody asked you to expand anything. Daan just pointed out a limitation without making a request; neither did I.

bluca commented:
> > > On Thu, Jan 05, 2023 at 02:49:12AM -0800, Daan De Meyer wrote: Unrelated to this PR but if we're going to require usage of this API for rootless nspawn it's going to be a pain since users will have to be running latest pid1 on the host to be able to use rootless nspawn. Building nspawn and repart and such from source is OK but having to run latest pid1 is a bit more complex.
> > > I personally think that requirement is ok. But if that's really something that's crucial to avoid we could do what I originally also proposed: make it a separate very very minimal namespace daemon binary that can be built alongside systemd-nspawn.
> > 
> > 
> > If we need systemd-backdoord.service we can always add it later and move the endpoint before the next release, this PR is already changing a fair few things, I don't really want to further expand it
> 
> That comment is weird. Nobody asked you to expand anything. Daan just pointed out a limitation without making a request; neither did I.

It will appear less weird if you check https://github.com/systemd/systemd/pull/25233#issuecomment-1366514673 - this PR was much smaller and contained at the beginning (still solving my problem), then after that I thought "what the heck, why not, how difficult could it be", so just making sure the merry-go-round doesn't start again ;-)

bluca commented:
> > > On Thu, Jan 05, 2023 at 02:49:12AM -0800, Daan De Meyer wrote: Unrelated to this PR but if we're going to require usage of this API for rootless nspawn it's going to be a pain since users will have to be running latest pid1 on the host to be able to use rootless nspawn. Building nspawn and repart and such from source is OK but having to run latest pid1 is a bit more complex.
> > > I personally think that requirement is ok. But if that's really something that's crucial to avoid we could do what I originally also proposed: make it a separate very very minimal namespace daemon binary that can be built alongside systemd-nspawn.
> > 
> > 
> > If we need systemd-backdoord.service we can always add it later and move the endpoint before the next release, this PR is already changing a fair few things, I don't really want to further expand it
> 
> That comment is weird. Nobody asked you to expand anything. Daan just pointed out a limitation without making a request; neither did I.

It will appear less weird if you check https://github.com/systemd/systemd/pull/25233#issuecomment-1366514673 - this PR was much smaller and contained at the beginning (still solving my problem), then after that I thought "what the heck, why not, how difficult could it be", so just making sure the merry-go-round doesn't start again ;-)

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,65 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+         * the kernel will refuse the operation, so don't fork directly in the new user namespace in that case.
+         * If we don't have to change the owner then it easier and we can let the clone set up the namespace
+         * instead via the apposite flag. */
+        if (owner_uid == UID_INVALID)
```
shouldn't this be `owner_uid == UID_INVALID || owner_gid == GID_INVALID`?

bluca replied:
I just figured you'd either change none, or just the uid, or both, but just the gid seems strange as a use case. But can change it

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,65 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+        if (r == 0) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
+                if (owner_uid != UID_INVALID) {
+                        if (owner_gid != GID_INVALID) {
```
why nest the gid conditionalization inside the uid conditionalzation?

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,65 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
-        if (r == 0)
-                /* Child. We do nothing here, just freeze until somebody kills us. */
+        if (r == 0) {
+                /* Child. Signal after setuid() and then just freeze until somebody kills us. */
```
doesn't matter, but maybe add as first thing here `pipe_fd[0] = safe_close(pipe_fd[0]);`. After all in the child we are never going to need the reading side of the pipe.

it doesn't matter of course, but does act as kind of a formalized documentation i think saying in which direction communication flows here.

bluca replied:
It's already closed, or rather, only pipe[1] is kept open across the fork by safe_fork_full

poettering commented on src/basic/namespace-util.c:
```diff
@@ -210,14 +213,65 @@ int userns_acquire(const char *uid_map, const char *gid_map) {
+                                return log_error_errno(r, "Failed to create user namespace: %m");
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Failed to create user namespace.");
+                }
+        }
```
hmm, i'd really send a `r = 0` integer from the child to the parent on success. Because otherwise, if the child dies abnormally it will look like everything is in order since we'l get POLLHUP then just like if everything was OK.

hence: the child should always send a message up, r == 0 in case of success, and something else on failure. and parent should always read the thing, and if it can't consider that an error

poettering commented on man/org.freedesktop.systemd1.xml:
```diff
@@ -265,6 +265,7 @@ node /org/freedesktop/systemd1 {
       LookupDynamicUserByUID(in  u uid,
                              out s name);
       GetDynamicUsers(out a(us) users);
+      GetIdentityMapping(out h namespace);
```
maybe GetIdentityUserNamespace()?

bluca replied:
that's a bit of a mouthful - that you get a namespace is encoded in the output parameter so to me it seems clearer this way

poettering commented on src/core/dbus-manager.c:
```diff
@@ -2783,6 +2783,62 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
+        if (r == 0)
+                return 1; /* No authorization for now, but the async polkit stuff will call us again when it has it */
+
+        r = sd_bus_query_sender_creds(message, SD_BUS_CREDS_UID|SD_BUS_CREDS_GID|SD_BUS_CREDS_AUGMENT, &creds);
```
SD_BUS_CREDS_AUGMENT stuff cannot be used for security decisions. we need to investigate this.

poettering commented on src/core/dbus-manager.c:
```diff
@@ -2783,44 +2783,83 @@ static int method_set_show_status(sd_bus_message *message, void *userdata, sd_bu
 
-        r = sd_bus_creds_get_gid(creds, &group);
+        /* User manager instances are allowed to bypass Polkit, and are always authorized as they will use this to set up
+         * sandboxing options for user units. Everything else goes through the Polkit auth step first. */
```
sorry, but we can't do that for security decision like that. pids can be recycled, and thus we can read stuff from /proc/$PID for decorative purposes, but not to make security decisions on. what's the rationale for trying to avoid PK lookups here?

bluca replied:
is there any other way to go from a dbus message to the systemd unit that it originates from?

poettering commented on src/core/org.freedesktop.systemd1.policy.in:
```diff
@@ -70,4 +70,14 @@
 
+        <action id="org.freedesktop.systemd1.acquire-namespace">
+                <description gettext-domain="systemd">Acquire pre-prepared namespace</description>
+                <message gettext-domain="systemd">Authentication is required to acquire pre-prepared namespace.</message>
```
hmm, why do we actually need to restrict access to identity mappings?

bluca replied:
sounds like we do

poettering commented:
so, hmm, @brauner why doesn't the kernel allow unpriv creation of identity userns mappings btw?  why wouldn't that be safe?

brauner commented:
> so, hmm, @brauner why doesn't the kernel allow unpriv creation of identity userns mappings btw? why wouldn't that be safe?

Your never supposed to gain more privileges than you had on the host. But when you own a userns you can always enter as userns root. IOW, you can always became UID 0 and it UID 0 is mapped to UID 0 on the host you're becoming real root. Sure, your caps are isolated but being real root already gives you a lot more power than just your own UID.

poettering commented on src/core/execute.c:
```diff
@@ -4226,6 +4226,44 @@ static int collect_open_file_fds(
+               context->n_bind_mounts > 0 ||
+               context->n_temporary_filesystems > 0 ||
+               context->root_directory ||
+               context->extension_directories ||
```
please check with !strv_is_empty()

poettering commented on src/core/execute.c:
```diff
@@ -4226,6 +4226,44 @@ static int collect_open_file_fds(
+               context->read_only_paths ||
+               context->inaccessible_paths ||
+               context->exec_paths ||
+               context->no_exec_paths;
```
these five the same

poettering commented on src/core/execute.c:
```diff
@@ -4226,6 +4226,44 @@ static int collect_open_file_fds(
+               context->protect_system != PROTECT_SYSTEM_NO ||
+               context->protect_home != PROTECT_HOME_NO ||
+               context->protect_kernel_tunables ||
+               context->protect_kernel_modules ||
```
i think we should not imply on these two, because this also enforces a seccomp filter, which we can do without PrivateUsers= and should entirely suffice.

poettering commented on src/core/execute.c:
```diff
@@ -4226,6 +4226,44 @@ static int collect_open_file_fds(
+               context->protect_home != PROTECT_HOME_NO ||
+               context->protect_kernel_tunables ||
+               context->protect_kernel_modules ||
+               context->protect_kernel_logs ||
```
this one similar actually

poettering commented on src/core/execute.c:
```diff
@@ -4226,6 +4226,44 @@ static int collect_open_file_fds(
+               context->protect_kernel_modules ||
+               context->protect_kernel_logs ||
+               context->protect_control_groups ||
+               context->protect_clock ||
```
and this one too

poettering commented:
So I think we should only turn on PrivateUsers= when otherwise the option is useless. some of the PrivateXYZ= settings result in *both* a seccomp filter and some namespacing changes, hence I think turning them on should not be enough to imply PrivateUsers=, because they are still somewhat effective without it (and oftentimes fully effective even)

bluca commented:
> So I think we should only turn on PrivateUsers= when otherwise the option is useless. some of the PrivateXYZ= settings result in _both_ a seccomp filter and some namespacing changes, hence I think turning them on should not be enough to imply PrivateUsers=, because they are still somewhat effective without it (and oftentimes fully effective even)

But if seccomp is disabled or unavailable, then we go back to silently working in a different way? The goal of this effort was to enforce consistency

bluca commented:
> > So I think we should only turn on PrivateUsers= when otherwise the option is useless. some of the PrivateXYZ= settings result in _both_ a seccomp filter and some namespacing changes, hence I think turning them on should not be enough to imply PrivateUsers=, because they are still somewhat effective without it (and oftentimes fully effective even)
> 
> But if seccomp is disabled or unavailable, then we go back to silently working in a different way? The goal of this effort was to enforce consistency

eg it means they'll have to be documented differently, as unavailable for user units in case seccomp is disabled, but still available for service units if seccomp is disabled. What's the advantage?

poettering commented:
> But if seccomp is disabled or unavailable, then we go back to silently working in a different way? The goal of this effort was to enforce consistency

Nah, most of these options have always been "best effort", as I mentioned many times: they'll do their thing if perms and kernel features allow it, but degrade gracefully if not. This is a good thing, as it means you can run the same unit file inside of a highly privileged, and featureful host – and inside a container with absolutely minimal available functionality and it will still work roughly the same, even though some of the lockdown options will not be applied where we cannot apply them.

So we *already* are behaving quite differently depending where we are invoked. This is not new.

I am still concerned that activating PrivateUsers= willy-nilly where it previously wasn't is a problem, since it modifies the execution environment so drastically (as suddenly a myriad of files will appear owned by "nobody"), and that's a compat issue I am sure. Now, there's a valid point to make that for settings that never worked in --user context at all this is a non-issue, since they couldn't have been used before in such a context, and hence we can just enable them here. But for things like ProtectKernelModules= (which users could already make use of in --user context before, where it would just enforce the seccomp part, but not the namespace part) I don't think we should just enable PrivateUsers= because if you do that anyone who assumed that previous, graceful  behaviour would now be affected by the big nobody flood of 2023.

poettering commented:
I mean your commit msg currently says this:

> Enabling these options when not running as root requires a user namespace, so implicitly enable PrivateUsers=. This has a side effect as it changes which users are visible to the unit.
However until now these options did not work at all for user units, and in practice just a handful of user units in Fedora, Debian and Ubuntu mistakenly used them (and they have been all fixed since).

But that's precsiely not true, i.e. the options that also do seccomp *did* work, and did have an effect in the status quo ante for --user instance: they just installed the seccomp thing and that's it.

bluca commented:
> > But if seccomp is disabled or unavailable, then we go back to silently working in a different way? The goal of this effort was to enforce consistency
> 
> Nah, most of these options have always been "best effort", as I mentioned many times: they'll do their thing if perms and kernel features allow it, but degrade gracefully if not. This is a good thing, as it means you can run the same unit file inside of a highly privileged, and featureful host – and inside a container with absolutely minimal available functionality and it will still work roughly the same, even though some of the lockdown options will not be applied where we cannot apply them.
> 
> So we _already_ are behaving quite differently depending where we are invoked. This is not new.

Sure, the problem is not the best effort part per se, but the difference in behaviour between system and user units. This confuses users and developers, and as a result very few user units ever make use of sandboxing, and users go pick other solutions instead of using them. By doing as you suggest, when seccomp is disabled/unavailable, on the exact same machine system units will still be able to implement those sandboxing while user units will not and instead will silently skip. This inconsistency is the problem I'm trying to solve.

> I am still concerned that activating PrivateUsers= willy-nilly where it previously wasn't is a problem, since it modifies the execution environment so drastically (as suddenly a myriad of files will appear owned by "nobody"), and that's a compat issue I am sure. Now, there's a valid point to make that for settings that never worked in --user context at all this is a non-issue, since they couldn't have been used before in such a context, and hence we can just enable them here. But for things like ProtectKernelModules= (which users could already make use of in --user context before, where it would just enforce the seccomp part, but not the namespace part) I don't think we should just enable PrivateUsers= because if you do that anyone who assumed that previous, graceful behaviour would now be affected by the big nobody flood of 2023.

I have not seen any user unit use those options in Debian, Ubuntu and Fedora. Nobody has replied to the email I sent last year to the mailing list. I have removed this PR from this milestone and moved it to the next. I could add a notice to v253's NEWS so that more people see it?

keszybz reviewed (changes requested):
I think we should merge this.

The only open issue is how to handle ProtectKernelModules= and other settings that are _partially_ implemented via seccomp. Lennart's approach gives slightly better backwards-compat, and Luca's approach (current patch) gives slightly more reasonable behaviour for users. I think we should be fine with the latter because of the preparatory work that was done: queries in distros and on the mailing list. The expectation is that that units that would be affected by this are very rare or nonexistent. Of course we will need to announce this in NEWS, etc, but I think we can get away with this. Let's merge this now, while there is still some time before the next release, to give it some time to "bake".

Needs a rebase and typo fixes.

keszybz commented:
CentOS CI (Arch Linux) fails deep in TEST-75-RESOLVED. But I don't think this is related, the unit is up and running when the failure happens.

keszybz commented:
jammy-ppc64el: TEST-29-PORTABLE:                   FAIL     (1809 s)
